File: src/ts/graph/graph_manifest.ts
/**
 * Fused graph-module manifest (`mathzig:graph` custom section + JSON sidecar).
 *
 * Mirrors `src/wasm/graph_manifest.zig`. Wire kinds reuse the same short names
 * as `mathzig:node` / `abi.WireKind` (number, matrix, …) — no second enum.
 *
 * Multi-value I/O contract (`abi: 1`):
 * - inputs then params (ordered); params may carry `default`
 * - outputs always a list of values; `out_mode` is `"table" | "named_exports"`
 * - table layout: [u32 count] × { [u32 kind_tag][f64 wire] } (names in manifest only)
 */

import { normalizePortKind } from "./value_transfer";
import { MAX_SOURCE_BYTES } from "./limits";
import { ManifestLoadError } from "./load_error";

/** Custom-section name (distinct from `mathzig:node` and `mathzig.abi`). */
export const GRAPH_MANIFEST_SECTION = "mathzig:graph";

/** Graph-manifest ABI version (independent of node-manifest). */
export const GRAPH_ABI_VERSION = 1;

/** Node custom-section name (shared with runner). */
export const NODE_MANIFEST_SECTION = "mathzig:node";

export type GraphOutMode = "table" | "named_exports";

export type GraphManifestPort = {
  name: string;
  kind: string;
  default?: number | null;
};

export type GraphManifestOutput = {
  name: string;
  kind: string;
  result_tag?: string;
  /** Present when `out_mode === "named_exports"`. */
  export?: string;
};

export type GraphManifestExport = {
  name: string;
  params: number;
};

/** Parsed `mathzig:graph` custom section / JSON sidecar. */
export type GraphManifest = {
  abi?: number;
  entry?: string;
  inputs: GraphManifestPort[];
  params: GraphManifestPort[];
  outputs: GraphManifestOutput[];
  out_mode?: GraphOutMode;
  exports?: GraphManifestExport[];
};

/**
 * Graph/node manifest validation failure (`phase: "manifest"`).
 * Extends the shared C3 load-error model.
 */
export class GraphManifestError extends ManifestLoadError {
  constructor(message: string, code = "InvalidManifest", path?: string) {
    super(code, message, path ? { position: { path } } : undefined);
    this.name = "GraphManifestError";
  }
}

/** Raw wasm-bytes scan for a named custom section (S3). */
export type CustomSectionScan =
  | { kind: "absent" }
  | { kind: "malformed_wasm"; reason: string }
  | { kind: "present"; payload: Uint8Array };

/**
 * Scan raw wasm bytes for a named custom section without instantiating.
 * Distinguishes malformed module bytes from an absent section (task-14 S3).
 */
export function scanCustomSectionBytes(
  wasmBytes: Uint8Array,
  sectionName: string,
): CustomSectionScan {
  if (wasmBytes.length < 8) {
    return { kind: "malformed_wasm", reason: "truncated header" };
  }
  if (
    wasmBytes[0] !== 0x00 ||
    wasmBytes[1] !== 0x61 ||
    wasmBytes[2] !== 0x73 ||
    wasmBytes[3] !== 0x6d
  ) {
    return { kind: "malformed_wasm", reason: "bad magic" };
  }
  let i = 8;
  const readLeb = (buf: Uint8Array, start: number): { value: number; next: number } | null => {
    let result = 0;
    let shift = 0;
    let j = start;
    while (j < buf.length) {
      const b = buf[j++]!;
      result |= (b & 0x7f) << shift;
      if ((b & 0x80) === 0) return { value: result >>> 0, next: j };
      shift += 7;
      if (shift > 35) return null;
    }
    return null;
  };
  while (i < wasmBytes.length) {
    const secId = wasmBytes[i++]!;
    const sizeRes = readLeb(wasmBytes, i);
    if (!sizeRes) return { kind: "malformed_wasm", reason: "bad section size LEB" };
    i = sizeRes.next;
    const payloadLen = sizeRes.value;
    if (i + payloadLen > wasmBytes.length) {
      return { kind: "malformed_wasm", reason: "truncated section payload" };
    }
    const payload = wasmBytes.subarray(i, i + payloadLen);
    i += payloadLen;
    if (secId !== 0) continue;
    const nameRes = readLeb(payload, 0);
    if (!nameRes) return { kind: "malformed_wasm", reason: "bad custom-section name LEB" };
    const nameLen = nameRes.value;
    if (nameRes.next + nameLen > payload.length) {
      return { kind: "malformed_wasm", reason: "truncated custom-section name" };
    }
    const name = new TextDecoder().decode(
      payload.subarray(nameRes.next, nameRes.next + nameLen),
    );
    if (name !== sectionName) continue;
    return { kind: "present", payload: payload.subarray(nameRes.next + nameLen) };
  }
  return { kind: "absent" };
}

const OUT_MODES = new Set<string>(["table", "named_exports"]);

function isObject(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

/** Reuse graph kind table; always throw GraphManifestError (not plain Error). */
function assertKnownKind(kind: string, label: string): void {
  try {
    normalizePortKind(kind);
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    throw new GraphManifestError(`${label}: ${msg}`);
  }
}

function assertPort(raw: unknown, label: string, allowDefault: boolean): GraphManifestPort {
  if (!isObject(raw)) throw new GraphManifestError(`${label} must be an object`);
  if (typeof raw.name !== "string" || raw.name.length === 0) {
    throw new GraphManifestError(`${label}.name must be a non-empty string`);
  }
  if (typeof raw.kind !== "string") {
    throw new GraphManifestError(`${label}.kind must be a string`);
  }
  assertKnownKind(raw.kind, `${label}.kind`);
  const port: GraphManifestPort = { name: raw.name, kind: raw.kind };
  if (allowDefault && "default" in raw) {
    const d = raw.default;
    if (d !== null && d !== undefined && typeof d !== "number") {
      throw new GraphManifestError(`${label}.default must be number | null`);
    }
    port.default = d as number | null | undefined;
  }
  return port;
}

function assertOutput(raw: unknown, index: number): GraphManifestOutput {
  if (!isObject(raw)) throw new GraphManifestError(`outputs[${index}] must be an object`);
  if (typeof raw.name !== "string" || raw.name.length === 0) {
    throw new GraphManifestError(`outputs[${index}].name must be a non-empty string`);
  }
  if (typeof raw.kind !== "string") {
    throw new GraphManifestError(`outputs[${index}].kind must be a string`);
  }
  assertKnownKind(raw.kind, `outputs[${index}].kind`);
  const out: GraphManifestOutput = { name: raw.name, kind: raw.kind };
  if (raw.result_tag !== undefined) {
    if (typeof raw.result_tag !== "string") {
      throw new GraphManifestError(`outputs[${index}].result_tag must be a string`);
    }
    out.result_tag = raw.result_tag;
  }
  if (raw.export !== undefined) {
    if (typeof raw.export !== "string") {
      throw new GraphManifestError(`outputs[${index}].export must be a string`);
    }
    out.export = raw.export;
  }
  return out;
}

function assertExport(raw: unknown, index: number): GraphManifestExport {
  if (!isObject(raw)) throw new GraphManifestError(`exports[${index}] must be an object`);
  if (typeof raw.name !== "string") {
    throw new GraphManifestError(`exports[${index}].name must be a string`);
  }
  if (typeof raw.params !== "number" || !Number.isFinite(raw.params)) {
    throw new GraphManifestError(`exports[${index}].params must be a finite number`);
  }
  return { name: raw.name, params: raw.params };
}

/**
 * Structural + kind validation for a parsed (or hand-built) graph manifest.
 * Throws `GraphManifestError` on empty outputs, unknown kinds, or bad shape.
 */
export function validateGraphManifest(raw: unknown): GraphManifest {
  if (!isObject(raw)) throw new GraphManifestError("manifest must be an object");

  const inputsRaw = raw.inputs ?? [];
  const paramsRaw = raw.params ?? [];
  const outputsRaw = raw.outputs;
  if (!Array.isArray(inputsRaw)) throw new GraphManifestError("inputs must be an array");
  if (!Array.isArray(paramsRaw)) throw new GraphManifestError("params must be an array");
  if (!Array.isArray(outputsRaw)) throw new GraphManifestError("outputs must be an array");
  if (outputsRaw.length === 0) throw new GraphManifestError("outputs must be non-empty");

  const inputs = inputsRaw.map((p, i) => assertPort(p, `inputs[${i}]`, false));
  const params = paramsRaw.map((p, i) => assertPort(p, `params[${i}]`, true));
  const outputs = outputsRaw.map((o, i) => assertOutput(o, i));

  // Duplicate port names within inputs, within params, or across both (task-14 S3).
  const seen = new Set<string>();
  for (const p of inputs) {
    if (seen.has(p.name)) {
      throw new GraphManifestError(`duplicate port name '${p.name}'`, "DuplicatePort", `inputs`);
    }
    seen.add(p.name);
  }
  for (const p of params) {
    if (seen.has(p.name)) {
      throw new GraphManifestError(`duplicate port name '${p.name}'`, "DuplicatePort", `params`);
    }
    seen.add(p.name);
  }

  let out_mode: GraphOutMode | undefined;
  if (raw.out_mode !== undefined) {
    if (typeof raw.out_mode !== "string" || !OUT_MODES.has(raw.out_mode)) {
      throw new GraphManifestError(`out_mode must be "table" | "named_exports"`);
    }
    out_mode = raw.out_mode as GraphOutMode;
  }

  let exportsList: GraphManifestExport[] | undefined;
  if (raw.exports !== undefined) {
    if (!Array.isArray(raw.exports)) throw new GraphManifestError("exports must be an array");
    exportsList = raw.exports.map((e, i) => assertExport(e, i));
  }

  const m: GraphManifest = {
    inputs,
    params,
    outputs,
  };
  if (typeof raw.abi === "number") m.abi = raw.abi;
  if (typeof raw.entry === "string") m.entry = raw.entry;
  if (out_mode !== undefined) m.out_mode = out_mode;
  if (exportsList !== undefined) m.exports = exportsList;
  return m;
}

/**
 * Read the `mathzig:graph` custom section from a compiled module.
 * Returns `null` when the section is absent (e.g. single-node or plain AOT).
 * When present, validates structure / kinds (throws `GraphManifestError` if bad).
 */
export function readGraphManifest(module: WebAssembly.Module): GraphManifest | null {
  const sections = WebAssembly.Module.customSections(module, GRAPH_MANIFEST_SECTION);
  if (sections.length === 0) return null;
  const text = new TextDecoder().decode(sections[0]);
  if (text.length > MAX_SOURCE_BYTES) {
    throw new GraphManifestError(
      `manifest JSON exceeds MAX_SOURCE_BYTES (${MAX_SOURCE_BYTES})`,
      "SourceTooLarge",
    );
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch {
    throw new GraphManifestError("manifest section is not valid JSON", "InvalidManifest");
  }
  return validateGraphManifest(parsed);
}

/**
 * S3 raw-bytes entry: scan wasm for `mathzig:graph`, distinguish absent vs
 * malformed wasm, then validate JSON payload.
 *
 * @returns `null` only when the section is **absent** on a well-formed module.
 * @throws `GraphManifestError` with code `MalformedWasm` for bad module bytes,
 *         or other codes for present-but-invalid manifests.
 */
export function readGraphManifestFromBytes(wasmBytes: Uint8Array): GraphManifest | null {
  const scan = scanCustomSectionBytes(wasmBytes, GRAPH_MANIFEST_SECTION);
  if (scan.kind === "absent") return null;
  if (scan.kind === "malformed_wasm") {
    throw new GraphManifestError(`malformed wasm: ${scan.reason}`, "MalformedWasm");
  }
  if (scan.payload.byteLength > MAX_SOURCE_BYTES) {
    throw new GraphManifestError(
      `manifest JSON exceeds MAX_SOURCE_BYTES (${MAX_SOURCE_BYTES})`,
      "SourceTooLarge",
    );
  }
  let text: string;
  try {
    text = new TextDecoder("utf-8", { fatal: true }).decode(scan.payload);
  } catch {
    throw new GraphManifestError("manifest section is not valid UTF-8", "InvalidUtf8");
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch {
    throw new GraphManifestError("manifest section is not valid JSON", "InvalidManifest");
  }
  return validateGraphManifest(parsed);
}

/**
 * S3 raw-bytes entry for `mathzig:node` (same absent/malformed distinction).
 * Returns the unvalidated JSON object when present; callers validate shape.
 */
export function readNodeManifestFromBytes(
  wasmBytes: Uint8Array,
): Record<string, unknown> | null {
  const scan = scanCustomSectionBytes(wasmBytes, NODE_MANIFEST_SECTION);
  if (scan.kind === "absent") return null;
  if (scan.kind === "malformed_wasm") {
    throw new GraphManifestError(`malformed wasm: ${scan.reason}`, "MalformedWasm");
  }
  if (scan.payload.byteLength > MAX_SOURCE_BYTES) {
    throw new GraphManifestError(
      `node manifest exceeds MAX_SOURCE_BYTES (${MAX_SOURCE_BYTES})`,
      "SourceTooLarge",
    );
  }
  let text: string;
  try {
    text = new TextDecoder("utf-8", { fatal: true }).decode(scan.payload);
  } catch {
    throw new GraphManifestError("node manifest section is not valid UTF-8", "InvalidUtf8");
  }
  try {
    const parsed = JSON.parse(text);
    if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) {
      throw new GraphManifestError("node manifest must be a JSON object", "ManifestNotObject");
    }
    return parsed as Record<string, unknown>;
  } catch (e) {
    if (e instanceof GraphManifestError) throw e;
    throw new GraphManifestError("node manifest section is not valid JSON", "InvalidManifest");
  }
}

File: src/wasm/graph_manifest.zig
//! Fused graph-module manifest (`mathzig:graph` custom section + JSON sidecar).
//!
//! Describes multi-input / multi-param / multi-output value ports of a single
//! AOT-fused wasm module so hosts can introspect I/O without re-parsing the
//! graph. Wire kinds reuse `abi.WireKind` via `node_manifest` helpers — no
//! second kind enum.
//!
//! # Normative multi-value I/O contract (`abi: 1`)
//!
//! - **Inputs then params**: ordered graph inputs, then ordered params.
//!   Each port is `{ name, kind }` with short kind names from
//!   `node_manifest.kindShortName` (number, boolean, matrix, complex, record,
//!   string, series, predicate, any). Params may carry `default: number | null`.
//! - **Outputs**: always a **list of values** (names live in the manifest).
//!   `out_mode` selects how the module surfaces them at the wasm boundary:
//!   - `"table"` (preferred default): after `tick`, host reads an output table
//!     from linear memory (base = i32 ptr returned by tick, or fixed offset):
//!     ```text
//!     [u32 count]
//!     repeated count times:
//!       [u32 kind_tag]   // abi.WireKind discriminant
//!       [f64 wire]       // number payload or ptr/handle as f64 bits
//!     ```
//!   - `"named_exports"`: each output may name a wasm export (`"export": "out_u"`).
//! - **Entry**: preferred single host entry is `entry` (default `"tick"`).
//! - **Kind ↔ result_tag**: reuse `node_manifest.kindToResultTag` /
//!   `parseKindName` (number→number, matrix_ptr→matrix, …).
//!
//! # Validation surface
//!
//! - Typed `GraphManifest` / `validate`: rejects empty `outputs` only (kinds are
//!   already `abi.WireKind` at this layer).
//! - JSON/host boundary: `parseGraphManifestJson` rejects unknown kind strings
//!   (`UnknownKind`), bad `out_mode` (`InvalidOutMode`), empty outputs, and
//!   structural garbage (`InvalidManifest`).
//! - `emitCustomSection` always runs `validate` so empty-output sections never
//!   land in a `.wasm`.
//!
//! Fused modules use this section instead of `mathzig:node` (single-node).
//! `mathzig.abi` may still be emitted for primary-export compiler metadata.
//!
//! Unit tests live in `tests/zig/backends/wasm/graph_manifest_test.zig` (single
//! source of truth; this module is not a test root).

const std = @import("std");
const abi = @import("abi.zig");
const list_writer = @import("list_writer.zig");
const node_manifest = @import("node_manifest.zig");
const module_mod = @import("module.zig");

/// Custom-section name (distinct from `mathzig:node` and `mathzig.abi`).
pub const CUSTOM_SECTION_NAME = "mathzig:graph";

/// Graph-manifest ABI version (independent of node-manifest / mathzig.abi).
pub const GRAPH_ABI_VERSION: u32 = 1;

/// Re-export port type shared with single-node manifests.
pub const Port = node_manifest.Port;

/// Kind helpers (no parallel enum — reuse node_manifest / abi.WireKind).
pub const parseKindName = node_manifest.parseKindName;
pub const kindShortName = node_manifest.kindShortName;
pub const kindToResultTag = node_manifest.kindToResultTag;

/// How multi-output values are exposed on the wasm boundary.
pub const OutMode = enum {
    /// Output table in linear memory (preferred default).
    table,
    /// Named exports `out_<name>` / explicit `export` field (scalar shortcut).
    named_exports,

    pub fn jsonName(self: OutMode) []const u8 {
        return switch (self) {
            .table => "table",
            .named_exports => "named_exports",
        };
    }

    pub fn parse(name: []const u8) ?OutMode {
        if (std.mem.eql(u8, name, "table")) return .table;
        if (std.mem.eql(u8, name, "named_exports")) return .named_exports;
        return null;
    }
};

/// One output value port (multi-out list; names only in manifest for table mode).
pub const GraphOutputPort = struct {
    name: []const u8,
    kind: abi.WireKind,
    /// `mathzig.abi` result_tag style; optional when derivable from `kind`.
    result_tag: ?[]const u8 = null,
    /// Wasm export name when `out_mode == named_exports` (e.g. `"out_u"`).
    export_name: ?[]const u8 = null,
};

/// Declared wasm export used by the fused module (params = f64 arity).
pub const ExportEntry = struct {
    name: []const u8,
    params: u32,
};

pub const GraphManifest = struct {
    /// Manifest ABI version (start at 1; independent of node-manifest).
    abi_version: u32 = GRAPH_ABI_VERSION,
    /// Preferred single host entry export name.
    entry: []const u8 = "tick",
    inputs: []const Port = &.{},
    params: []const Port = &.{},
    outputs: []const GraphOutputPort = &.{},
    out_mode: OutMode = .table,
    exports: []const ExportEntry = &.{},
};

/// Errors from `validate` on an already-typed manifest.
pub const ValidateError = error{
    EmptyOutputs,
};

/// Errors from JSON / host-boundary parse (`parseGraphManifestJson`).
pub const ParseError = error{
    InvalidManifest,
    EmptyOutputs,
    UnknownKind,
    InvalidOutMode,
} || error{OutOfMemory};

/// Structural checks for a typed `GraphManifest`.
///
/// Kinds are already `abi.WireKind` here, so unknown-kind rejection happens in
/// `parseGraphManifestJson` (string → WireKind) — not in this function.
pub fn validate(manifest: GraphManifest) ValidateError!void {
    if (manifest.outputs.len == 0) return error.EmptyOutputs;
}

/// Parse a `mathzig:graph` JSON document into an arena-owned `GraphManifest`.
/// Port name / result_tag / entry / export name slices live in `arena`.
pub fn parseGraphManifestJson(arena: std.mem.Allocator, json_text: []const u8) ParseError!GraphManifest {
    const root = std.json.parseFromSliceLeaky(std.json.Value, arena, json_text, .{}) catch return error.InvalidManifest;
    if (root != .object) return error.InvalidManifest;
    const obj = root.object;

    const abi_version: u32 = if (obj.get("abi")) |a| blk: {
        break :blk switch (a) {
            .integer => |i| @intCast(i),
            .float => |f| @intFromFloat(f),
            else => return error.InvalidManifest,
        };
    } else GRAPH_ABI_VERSION;

    const entry: []const u8 = if (obj.get("entry")) |e|
        switch (e) {
            .string => |s| s,
            else => return error.InvalidManifest,
        }
    else
        "tick";

    var inputs: std.ArrayListUnmanaged(Port) = .empty;
    if (obj.get("inputs")) |ins| {
        if (ins != .array) return error.InvalidManifest;
        for (ins.array.items) |item| {
            try inputs.append(arena, try parsePort(item, false));
        }
    }

    var params: std.ArrayListUnmanaged(Port) = .empty;
    if (obj.get("params")) |par| {
        if (par != .array) return error.InvalidManifest;
        for (par.array.items) |item| {
            try params.append(arena, try parsePort(item, true));
        }
    }

    var outputs: std.ArrayListUnmanaged(GraphOutputPort) = .empty;
    if (obj.get("outputs")) |outs| {
        if (outs != .array) return error.InvalidManifest;
        for (outs.array.items) |item| {
            try outputs.append(arena, try parseOutput(item));
        }
    }

    const out_mode: OutMode = if (obj.get("out_mode")) |om| blk: {
        if (om != .string) return error.InvalidManifest;
        break :blk OutMode.parse(om.string) orelse return error.InvalidOutMode;
    } else .table;

    var exports_list: std.ArrayListUnmanaged(ExportEntry) = .empty;
    if (obj.get("exports")) |exs| {
        if (exs != .array) return error.InvalidManifest;
        for (exs.array.items) |item| {
            try exports_list.append(arena, try parseExport(item));
        }
    }

    const m = GraphManifest{
        .abi_version = abi_version,
        .entry = entry,
        .inputs = try inputs.toOwnedSlice(arena),
        .params = try params.toOwnedSlice(arena),
        .outputs = try outputs.toOwnedSlice(arena),
        .out_mode = out_mode,
        .exports = try exports_list.toOwnedSlice(arena),
    };
    try validate(m);
    return m;
}

fn parsePort(item: std.json.Value, allow_default: bool) ParseError!Port {
    if (item != .object) return error.InvalidManifest;
    const name = switch (item.object.get("name") orelse return error.InvalidManifest) {
        .string => |s| s,
        else => return error.InvalidManifest,
    };
    const kind_s = switch (item.object.get("kind") orelse return error.InvalidManifest) {
        .string => |s| s,
        else => return error.InvalidManifest,
    };
    const kind = parseKindName(kind_s) orelse return error.UnknownKind;
    var default: ?f64 = null;
    if (allow_default) {
        if (item.object.get("default")) |d| {
            default = switch (d) {
                .null => null,
                .integer => |i| @floatFromInt(i),
                .float => |f| f,
                .number_string => |s| std.fmt.parseFloat(f64, s) catch return error.InvalidManifest,
                else => return error.InvalidManifest,
            };
        }
    }
    return .{ .name = name, .kind = kind, .default = default };
}

fn parseOutput(item: std.json.Value) ParseError!GraphOutputPort {
    if (item != .object) return error.InvalidManifest;
    const name = switch (item.object.get("name") orelse return error.InvalidManifest) {
        .string => |s| s,
        else => return error.InvalidManifest,
    };
    const kind_s = switch (item.object.get("kind") orelse return error.InvalidManifest) {
        .string => |s| s,
        else => return error.InvalidManifest,
    };
    const kind = parseKindName(kind_s) orelse return error.UnknownKind;
    var result_tag: ?[]const u8 = null;
    if (item.object.get("result_tag")) |t| {
        if (t != .string) return error.InvalidManifest;
        result_tag = t.string;
    }
    var export_name: ?[]const u8 = null;
    if (item.object.get("export")) |ex| {
        if (ex != .string) return error.InvalidManifest;
        export_name = ex.string;
    }
    return .{
        .name = name,
        .kind = kind,
        .result_tag = result_tag,
        .export_name = export_name,
    };
}

fn parseExport(item: std.json.Value) ParseError!ExportEntry {
    if (item != .object) return error.InvalidManifest;
    const name = switch (item.object.get("name") orelse return error.InvalidManifest) {
        .string => |s| s,
        else => return error.InvalidManifest,
    };
    const params: u32 = switch (item.object.get("params") orelse return error.InvalidManifest) {
        .integer => |i| @intCast(i),
        .float => |f| @intFromFloat(f),
        else => return error.InvalidManifest,
    };
    return .{ .name = name, .params = params };
}

/// Serialize a graph manifest as compact JSON (custom section + sidecar).
/// Does not validate; call `validate` or use `emitCustomSection` for emit guards.
pub fn writeJson(manifest: GraphManifest, writer: anytype) !void {
    try writer.print("{{\"abi\":{d},\"entry\":\"", .{manifest.abi_version});
    try writeEscaped(writer, manifest.entry);
    try writer.writeAll("\",\"inputs\":[");
    for (manifest.inputs, 0..) |port, i| {
        if (i > 0) try writer.writeAll(",");
        try writePortJson(writer, port, false);
    }
    try writer.writeAll("],\"params\":[");
    for (manifest.params, 0..) |port, i| {
        if (i > 0) try writer.writeAll(",");
        try writePortJson(writer, port, true);
    }
    try writer.writeAll("],\"outputs\":[");
    for (manifest.outputs, 0..) |out, i| {
        if (i > 0) try writer.writeAll(",");
        try writeOutputJson(writer, out);
    }
    try writer.writeAll("],\"out_mode\":\"");
    try writer.writeAll(manifest.out_mode.jsonName());
    try writer.writeAll("\",\"exports\":[");
    for (manifest.exports, 0..) |ex, i| {
        if (i > 0) try writer.writeAll(",");
        try writer.writeAll("{\"name\":\"");
        try writeEscaped(writer, ex.name);
        try writer.print("\",\"params\":{d}}}", .{ex.params});
    }
    try writer.writeAll("]}");
}

fn writePortJson(writer: anytype, port: Port, include_default: bool) !void {
    try writer.writeAll("{\"name\":\"");
    try writeEscaped(writer, port.name);
    try writer.writeAll("\",\"kind\":\"");
    try writer.writeAll(kindShortName(port.kind));
    try writer.writeAll("\"");
    if (include_default) {
        if (port.default) |d| {
            try writer.print(",\"default\":{d}", .{d});
        } else {
            try writer.writeAll(",\"default\":null");
        }
    }
    try writer.writeAll("}");
}

fn writeOutputJson(writer: anytype, out: GraphOutputPort) !void {
    try writer.writeAll("{\"name\":\"");
    try writeEscaped(writer, out.name);
    try writer.writeAll("\",\"kind\":\"");
    try writer.writeAll(kindShortName(out.kind));
    try writer.writeAll("\",\"result_tag\":\"");
    const tag = out.result_tag orelse kindToResultTag(out.kind);
    try writeEscaped(writer, tag);
    try writer.writeAll("\"");
    if (out.export_name) |ex| {
        try writer.writeAll(",\"export\":\"");
        try writeEscaped(writer, ex);
        try writer.writeAll("\"");
    }
    try writer.writeAll("}");
}

fn writeEscaped(writer: anytype, text: []const u8) !void {
    for (text) |ch| {
        switch (ch) {
            '"' => try writer.writeAll("\\\""),
            '\\' => try writer.writeAll("\\\\"),
            '\n' => try writer.writeAll("\\n"),
            '\r' => try writer.writeAll("\\r"),
            '\t' => try writer.writeAll("\\t"),
            else => if (ch < 0x20)
                try writer.print("\\u{x:0>4}", .{ch})
            else
                try writer.writeAll(&.{ch}),
        }
    }
}

/// Allocate JSON bytes for a manifest (caller frees). Does not validate.
pub fn toJsonAlloc(allocator: std.mem.Allocator, manifest: GraphManifest) ![]u8 {
    var buf = std.ArrayListUnmanaged(u8).empty;
    errdefer buf.deinit(allocator);
    const w = list_writer.unmanagedByteWriter(&buf, allocator);
    try writeJson(manifest, w);
    return try buf.toOwnedSlice(allocator);
}

/// Append a `mathzig:graph` custom section to a WasmModule.
/// Runs `validate` first so empty-output manifests never land in `.wasm`.
pub fn emitCustomSection(module: anytype, allocator: std.mem.Allocator, manifest: GraphManifest) !void {
    try validate(manifest);
    const json = try toJsonAlloc(allocator, manifest);
    defer allocator.free(json);
    try module.addCustomSection(CUSTOM_SECTION_NAME, json);
}

/// Typed scan result (task-14 / C3): absent vs malformed_wasm vs present.
pub const SectionScan = module_mod.CustomSectionScan;

/// Scan wasm bytes for a `mathzig:graph` custom section (typed).
pub fn scanWasmGraphSectionResult(wasm_bytes: []const u8) SectionScan {
    return module_mod.scanCustomSectionResult(wasm_bytes, CUSTOM_SECTION_NAME);
}

/// Legacy: payload or null when absent **or** malformed.
pub fn scanWasmGraphSection(wasm_bytes: []const u8) ?[]const u8 {
    return switch (scanWasmGraphSectionResult(wasm_bytes)) {
        .present => |p| p,
        .absent, .malformed_wasm => null,
    };
}
